net/libfab/libfab.c

```diff
@@ -0,0 +1,2829 @@
+ */
+static int libfab_ep_find(struct m0_net_transfer_mc *tm, const char *name,
+			  struct m0_fab__ep_name *epn,
+			  struct m0_net_end_point **epp)
+{
+	struct m0_net_end_point  *net;
+	struct m0_fab__ep        *ep;
+	struct m0_fab__active_ep *aep;
+	struct m0_fab__tm        *ma;
+	char                      ep_str[LIBFAB_ADDR_STRLEN_MAX] = {'\0'};
+	char                     *wc = NULL;
+	bool                      found = false;
+	int                       rc = 0;
+
+	M0_ASSERT(libfab_tm_is_locked(tm->ntm_xprt_private));
+	m0_tl_for(m0_nep, &tm->ntm_end_points, net) {
+		ep = libfab_ep_net(net);
+		if (libfab_ep_cmp(ep, name, epn)) {
+			*epp = &ep->fep_nep;
+			found = true;
+			libfab_ep_get(ep);
+			break;
+		}
+	} m0_tl_endfor;
+
+	if (found == false) {
+		if (name != NULL)
+			rc = libfab_ep_create(tm, name, epn, epp);
+		else {
+			M0_ASSERT(epn != NULL);
+			M0_ASSERT((strlen(epn->fen_addr) + strlen(epn->fen_port) 
+				  + 8) < LIBFAB_ADDR_STRLEN_MAX);
+			sprintf(ep_str, "libfab:%s:%s", epn->fen_addr,
+				epn->fen_port);
+			rc = libfab_ep_create(tm, ep_str, epn, epp);
+		}
+	} else {
+		if (name != NULL && epn != NULL) {
+			wc = strchr(name,'*');
+			if (wc != NULL && 
+			    strcmp(ep->fep_name.fen_port, epn->fen_port) != 0) {
+				strcpy(ep->fep_name.fen_addr, epn->fen_addr);
+				strcpy(ep->fep_name.fen_port, epn->fen_port);
+				aep = libfab_aep_get(ep);
+				ma = tm->ntm_xprt_private;
+				if (aep->aep_tx_state == FAB_CONNECTED) {
+					libfab_txep_init(aep, ma);
+				}
+			}
+		}
+	}
+
+	return M0_RC(rc);
+}
+
+/**
+ * Used to create new active endpoint
+ */
+static int libfab_ep_create(struct m0_net_transfer_mc *tm, const char *name,
+			    struct m0_fab__ep_name *epn,
+			    struct m0_net_end_point **epp)
+{
+	struct m0_fab__tm *ma = tm->ntm_xprt_private;
+	struct m0_fab__ep *ep = NULL;
+	char              *wc;
+	int                rc;
+
+	M0_ENTRY("name=%s", name);
+	M0_PRE(name != NULL);
+
+	M0_ALLOC_PTR(ep);
+	if (ep == NULL)
+		return M0_ERR(-ENOMEM);
+
+	M0_ALLOC_PTR(ep->fep_aep);
+	if (ep->fep_aep == NULL)
+		return M0_ERR(-ENOMEM);
+
+	ep->fep_listen = NULL;
+
+	rc = libfab_ep_addr_decode(ep, name);
+	if (rc != FI_SUCCESS) {
+		libfab_aep_param_free(ep->fep_aep, ma);
+		return M0_RC(rc);
+	}
+
+	wc = strchr(name, '*');
+	if (epn != NULL && wc != NULL) {
+		strcpy(ep->fep_name.fen_addr, epn->fen_addr);
+		strcpy(ep->fep_name.fen_port, epn->fen_port);
+	}
+
+	rc = libfab_active_ep_create(ep, ma);
+	if (rc != FI_SUCCESS) {
+		return M0_RC(rc);
+	}
+
+	fab_sndbuf_tlist_init(&ep->fep_sndbuf);
+	*epp = &ep->fep_nep;
+	return M0_RC(rc);
+}
+
+/**
+ * Init resources for a transfer machine
+ */
+static int libfab_tm_res_init(struct m0_fab__tm *tm)
+{
+	struct m0_fab__fab        *fab;
+	struct m0_fab__passive_ep *pep;
+	struct fi_cq_attr          cq_attr;
+	int                        rc = 0;
+	
+	M0_PRE(tm != NULL);
+
+	pep = tm->ftm_pep->fep_listen;
+	fab = tm->ftm_fab;
+	memset(&cq_attr, 0, sizeof(cq_attr));
```

Comment:
We can either:
```
M0_SET0(&cq_attr);
```
or declare the variable as something like this:
```
struct fi_cq_attr cq_attr = { 0 };
```

Reply:
Corrected